ambry-store/src/main/java/com.github.ambry.store/HardDeleter.java

```diff
@@ -172,8 +174,6 @@ void pause() throws StoreException, IOException {
       if (paused.compareAndSet(false, true)) {
         logger.info("HardDelete thread has been paused ");
         index.persistIndex();
-        pruneHardDeleteRecoveryRange();
-        persistCleanupToken();
```

Comment:
we don't need to persistCleanupToken in pause method, since there are only two cases we have to persist cleanup token to disk. One is when the startTokenSafeToPersist is changed, which is dealt with postLogFlush method. The other one is when the hardDeleteRecoveryRange is changed, which is dealt with performHardDelete.

Reply:
Does this mean when `pause` method gets the undelete lock, the `performHardDelete` should be complete and token should have been persisted already?

Reply:
If so, `IOException` in method signature can be removed

Reply:
that's right. Persisting recovery information to disk is not a periodical action, it's triggered on demand. When 'startTokenSafeToPersist' changes, we should write it to file. And this is taking place in postLogFlush method. When the hardDeleteRecoveryRange or endToken changes, we should write it to file. And this is taking place in performHardDelete method. So there is no need to persist it here or in shutdown.